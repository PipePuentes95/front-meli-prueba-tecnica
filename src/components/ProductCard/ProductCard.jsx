import React from 'react';
import { useNavigate } from 'react-router-dom';

import './ProductCard.scss';
import iconFreeShipping from '../../assets/icon-free-shipping.png';

const ProductCard = ({
  id,
  picture,
  price,
  free_shipping,
  state_name,
  title,
}) => {

  const navigate = useNavigate();

  const handlerGoToProductDetail = () => {
    navigate(`/items/${ id }`);
  }

  const formatAmount = (value) => {
    let format = Math.trunc(value);
    format = format.toString();
    return format.replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1.');;
  }

  const formatDecimals = (value) => {
   return value.toFixed(2).replace('0.', '');
  }

  return (
    <div className='product-card' onClick={ handlerGoToProductDetail }>
      <div className="pc-img">
        <div className="pc-img__wrapper">
          <img src={ picture } alt="img test" width="180" height="180" />
        </div>
      </div>
      <div className="pc-content">
        <div className="pc-content-header">
          <div className="pc-price">
            <p className="price">
              $ { formatAmount(price.amount)  }
              { price.decimals !== 0 &&  <span className='price__decimals'>{ formatDecimals(price.decimals) }</span> }
            </p>
            { free_shipping && 
              <img className='free-shipping' src={ iconFreeShipping } alt="icon free shipping" width="18" height="18" /> 
            }
            
          </div>
          <div className="pc-location">
            <span className='location'>{ state_name }</span>
          </div>
        </div>
        <div className="pc-content-body">
          <p className='pc-title'>{ title }</p>
        </div>
      </div>
    </div>
  );
};

export default ProductCard;
