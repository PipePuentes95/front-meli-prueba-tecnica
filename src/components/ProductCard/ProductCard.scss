@import '../../styles/variables';

.product-card {
  width: 100%;
  position: relative;

  display: flex;

  background-color: $white;

  cursor: pointer;

  &:not(:last-of-type)::before {
    position: absolute;
    bottom: 0;
    left: 50%;
    
    width: 95%;
    height: 2px;
    
    background: $grey-4;

    content: '';

    transform: translateX(-50%);
  }

  .pc-img {
    padding: 16px;
    
    @media(max-width: 576px) {
      padding: 8px;
    }

    &__wrapper {
      width: 120px;
      height: 120px;

      border-radius: 4px;

      overflow: hidden;
      
      img {
        width: 100%;
        height: 100%;
  
        object-fit: contain;
        object-position: center;
      }
    }
  }

  .pc-content {
    width: 100%;
    max-width: 700px;

    padding: 16px 8px 16px 0;

    .pc-content-header {
      display: flex;
      flex-wrap: wrap;
      align-items: flex-end;
      margin-bottom: 32px;

      @media(max-width: 576px) {
        margin-bottom: 14px;
      }

      .pc-price {
        display: flex;
        align-items: center;

        .price {
          display: flex;
          align-items: flex-start;
          padding-right: 12px;

          font-size: 18px;
          color: $grey-1;

          &__decimals {
            margin-top: 2px;
            margin-left: 4px;
            font-size: 15px;
          }
        }

        .free-shipping {
          width: 18px;
          height: 18px;
        }
      }

      .pc-location {
        @media(min-width: 577px) {
          margin-left: auto;
        }

        width: 150px;

        .location {
          font-size: 12px;
          color: $grey-2;
        }
      }
    }

    .pc-content-body {
      max-width: 400px;

      .pc-title {
        font-size: 18px;
        color: $grey-1;

        @media(max-width: 576px) {
          font-size: 14px;
        }
      }
    }
  }
}
