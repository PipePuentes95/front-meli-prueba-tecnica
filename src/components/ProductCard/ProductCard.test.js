import { render, fireEvent, screen } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import ProductCard from './ProductCard';

const id = "MLA1110584659";
const picture = "http://http2.mlstatic.com/D_947304-MLA48082618871_102021-O.jpg";
const price = {
  "currency": "ARS",
  "amount": 10499,
  "decimals": 0
};
const free_shipping = true;
const state_name = "Buenos Aires";
const title = "Zapatillas adidas Coreracer 2771 Mark";

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => (jest.fn())
})); 

it('should take a snapshot', () => {
  const pathname = createMemoryHistory();
  const { asFragment } = render (
    <Router location={ pathname.location }>
      <ProductCard
        key={ id }
        id={ id }
        title={ title }
        price={ price }
        picture={ picture }
        free_shipping={ free_shipping }
        state_name={ state_name } />
    </Router>
  )

  expect(asFragment(<ProductCard />)).toMatchSnapshot();
});

it('pathname should be /', () => {
  const history = createMemoryHistory({ initialEntries: ['/'] });
  
  render (
    <Router location={ history.location }>
      <ProductCard
        key={ id }
        id={ id }
        title={ title }
        price={ price }
        picture={ picture }
        free_shipping={ free_shipping }
        state_name={ state_name } />
    </Router>
  )

  expect(history.location.pathname).toBe('/');

  fireEvent.click(screen.getByTestId('go-to-product'));

  // expect(history.location.pathname).toBe('/items/123');
})
