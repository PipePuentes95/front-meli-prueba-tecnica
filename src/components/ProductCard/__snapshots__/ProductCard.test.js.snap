// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="product-card"
    data-testid="go-to-product"
  >
    <div
      class="pc-img"
    >
      <div
        class="pc-img__wrapper"
      >
        <img
          alt="img test"
          height="180"
          src="http://http2.mlstatic.com/D_947304-MLA48082618871_102021-O.jpg"
          width="180"
        />
      </div>
    </div>
    <div
      class="pc-content"
    >
      <div
        class="pc-content-header"
      >
        <div
          class="pc-price"
        >
          <p
            class="price"
          >
            $ 10.499
          </p>
          <img
            alt="icon free shipping"
            class="free-shipping"
            height="18"
            src="icon-free-shipping.png"
            width="18"
          />
        </div>
        <div
          class="pc-location"
        >
          <span
            class="location"
          >
            Buenos Aires
          </span>
        </div>
      </div>
      <div
        class="pc-content-body"
      >
        <p
          class="pc-title"
        >
          Zapatillas adidas Coreracer 2771 Mark
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
